### PR TITLE
Remove usage of mappers for applications on frontend #167210415

### PR DESCRIPTION
**Pull request by akegan**
[ticket](https://www.pivotaltracker.com/story/show/167210415)

### CODE REVIEW

File: app/controllers/api/v1/short_form_controller.rb [hunk 1]
```diff
@@ -6,159 +6,35 @@ class Api::V1::ShortFormController < ApiController
...
-      end
-
-      render json: { application: application }
-    else
-      render status: 422, json: { errors: short_form_validator.errors.full_messages || 'Unknown Error' }
+    # short_form_validator = ShortFormValidator.new(application_api_params)
+
+
+    # if short_form_validator.valid?
+    # FIXME: validatate / permit on the params
```
Comment (akegan) [inline app/controllers/api/v1/short_form_controller.rb]:
```suggestion
```

This has already been fixed. The ShortFormValidator wasn't doing anything

File: app/controllers/api/v1/short_form_controller.rb [hunk 2]
```diff
@@ -6,159 +6,35 @@ class Api::V1::ShortFormController < ApiController
...
+    custom_api_application = Force::Application.from_domain(application_api_params).to_custom_api
+    puts 'Custom api application', custom_api_application.to_json
+    application = custom_api_application_service.submit(custom_api_application)
+    logger.debug "application submit response: #{application}"
+    puts "application submit response: #{application}"
```
Comment (akegan) [inline app/controllers/api/v1/short_form_controller.rb]:
```suggestion
```

File: app/controllers/api/v1/short_form_controller.rb [hunk 3]
```diff
@@ -6,159 +6,35 @@ class Api::V1::ShortFormController < ApiController
...
+    # if submitting a supplemental application we will re-fetch updated data from salesforce
+    if params[:supplemental]
+      application = soql_application_service.application(application[:id] || application_api_params[:id], { include_lease: false })
+      logger.debug "updated application: #{application}"
     end
+
+    render json: { application: application }
+    # else
```
Comment (akegan) [inline app/controllers/api/v1/short_form_controller.rb:22]: this still needs to be cleaned up/fixed

File: app/javascript/components/applications/ApplicationEditPage.js [hunk 4]
```diff
@@ -8,6 +8,7 @@ import { saveApplication } from './actions'
 
 const ApplicationEditPageForm = ({ listing, application, editPage, lendingInstitutions }) => {
   const saveEditApplication = async (submitType, submittedValues, listing, editPage) => {
+    submittedValues.listing_id = listing.id
```
Comment (akegan) [inline app/javascript/components/applications/ApplicationEditPage.js:11]: This is sort of a major problem still left with these updates: what to do with the listing Id. We have to send it, and in some cases it's stored in the application object as:
"listing": {
"id": "blah"
}

and in some cases it's stored as listing_id. I've been trying to standardize around listing_id because that is the only piece that comes with the application and is saved with the application. For other listing information, I'd rather keep it out of the listing object.

File: app/javascript/components/mappers/soqlToDomain/application.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ import { mapListing } from './listing'
 import { mapUser } from './user'
 import { mapShape, mapList } from '../utils'
 
-const parseList = text => split(text, ';')
+const parseList = text => { return (text ? split(text, ';') : []) }
```
Comment (akegan) [inline app/javascript/components/mappers/soqlToDomain/application.js:9]: there was actually a bug where ADA info wasn't showing and saving correctly on supplemental applications

File: app/models/force/application.rb [hunk 6]
```diff
@@ -35,8 +38,10 @@ class Application < Force::ObjectBase
       { custom_api: 'householdVouchersSubsidies', domain: 'housing_voucher_or_subsidy', salesforce: 'Housing_Voucher_or_Subsidy' },
       { custom_api: 'housingCounselingAgency', domain: '', salesforce: 'Which_One' },
       { custom_api: 'id', domain: 'id', salesforce: 'Id' },
+      { custom_api: 'isFirstTimeHomebuyer', domain: 'is_first_time_homebuyer', salesforce: '?' },
```
Comment (akegan) [inline app/models/force/application.rb:41]: I left ? as the value for Salesforce in a lot of cases, focusing on domain to customAPI in this PR and just trying not to break Salesforce to domain translation that was already happening.

File: app/models/force/application.rb [hunk 7]
```diff
@@ -58,6 +64,11 @@ class Application < Force::ObjectBase
       { custom_api: 'hasCompletedHomebuyerEducation', domain: 'has_completed_homebuyer_education', salesforce: 'Has_Completed_Homebuyer_Education' },
       { custom_api: 'hasLoanPreapproval', domain: 'has_loan_preapproval', salesforce: 'Has_Loan_Pre_approval' },
       { custom_api: 'lendingAgent', domain: 'lending_agent', salesforce: 'Lending_Agent' },
+      { custom_api: 'shortFormPreferences', domain: 'preferences', salesforce: '?', object: Force::Preference},
```
Comment (akegan) [inline app/models/force/application.rb]: It would be nice to use this info about "object" to do nested mapping, but since some of them are a list of the object and some of them aren't, and every nested object has a bit of a special case I haven't yet

File: app/models/force/application.rb [hunk 8]
```diff
@@ -58,6 +64,11 @@ class Application < Force::ObjectBase
...
+      { custom_api: 'primaryApplicant', domain: 'applicant', salesforce: '?', object: Force::ApplicationMember},
+      { custom_api: 'alternateContact', domain: 'alternate_contact', salesforce: '?', object: Force::ApplicationMember},
+      { custom_api: 'householdMembers', domain: 'household_members', salesforce: '?', object: Force::ApplicationMember},
+      { custom_api: '', domain: 'demographics', salesforce: '?', object: Force::Demographics},
```
Comment (akegan) [inline app/models/force/application.rb]: demographics are weird because we store them on the application object in the paper application, but we have to send them to salesforce on the primary applicant object

File: app/models/force/application_member.rb [hunk 9]
```diff
@@ -32,8 +32,24 @@ class ApplicationMember < Force::ObjectBase
...
+      { custom_api: 'sexualOrientation', domain: 'sexual_orientation', salesforce: '?'},
+      { custom_api: 'race', domain: 'race', salesforce: '?'},
+      { custom_api: 'genderOther', domain: 'gender_other', salesforce: '?'},
+      { custom_api: 'sexualOrientationOther', domain: 'sexual_orientation_other', salesforce: '?'}
     ].freeze
 
+    def to_custom_api
+      custom_api_fields = super
+      # Convert array date into string
+      if @fields.domain and @fields.domain['date_of_birth']
```
Comment (akegan) [inline app/models/force/application_member.rb:47]: I think we could do something fancy with converting dates using the "object" key in the field definitions, but that also seems out of scope for this PR

File: app/models/force/object_base.rb [hunk 10]
```diff
@@ -73,16 +73,35 @@ def currency_to_float(field_name, salesforce_fields)
...
     end
 
+    # TODO: Automate this conversion for date objects.
     def self.date_to_salesforce(domain_date)
       return nil unless !domain_date.blank? && %i[year month day].all? {|s| domain_date.key? s}
       lease_date = Date.new(domain_date[:year].to_i, domain_date[:month].to_i, domain_date[:day].to_i)
       lease_date.strftime('%F')
     end
+
+    def self.get_domain_keys(obj = self)
```
Comment (akegan) [inline app/models/force/object_base.rb:95]: I created this so that we could automate which fields we permit in the short form controller

File: spec/fixtures/model/force/application/paper_app_domain.json [hunk 11]
```diff
@@ -0,0 +1,114 @@
+{
```
Comment (akegan) [inline spec/fixtures/model/force/application/paper_app_domain.json:1]: I manually created this json by filling out every field in a paper application on the frontend

File: spec/javascript/components/supplemental_application/__snapshots__/SupplementalApplicationPage.test.js.snap [hunk 12]
```diff
@@ -5908,9 +5890,7 @@ exports[`SupplementalApplicationPage should render the status update dropdown bu
                                             "flagged_applications": Array [],
                                             "general_lottery": true,
                                             "general_lottery_rank": null,
-                                            "has_ada_priorities_selected": Object {
-                                              "": true,
```
Comment (akegan) [inline spec/javascript/components/supplemental_application/__snapshots__/SupplementalApplicationPage.test.js.snap:5912]: This is what the bug was with ADA values on the supp app. this extra empty string key was floating around

File: spec/model/force/application_spec.rb [hunk 13]
```diff
@@ -3,14 +3,16 @@
 require 'rails_helper'
 
 RSpec.describe Force::Application do
+  let(:paper_app_custom_api) { fixture('model/force/application/paper_app_custom_api.json') }
```
Comment (akegan) [inline spec/model/force/application_spec.rb:6]: I'd like to add even more tests here to make sure the functions are really well covered. We have had so many bugs in the past becuase of missing fields or things being mis-translated. This is our chance to be really explicit about what we want to happen when we map from domain to customAPI

File: spec/model/force/object_base_spec.rb [hunk 14]
```diff
@@ -31,6 +31,13 @@
       expect(Force::ObjectBase.date_to_salesforce({})).to eq(nil)
     end
   end
+
+  describe '#get_domain_keys' do
+    it 'generates a list of all of the permitted keys' do
+      result = Force::Application.get_domain_keys()
+      expect(result).to eq 'something'
```
Comment (akegan) [inline spec/model/force/object_base_spec.rb]: obviously an unfinished test 😂

**Review by akegan — COMMENTED** [review]
Some additional context for @software-project.

File: app/controllers/api/v1/short_form_controller.rb (hunk 3 above)
Comment (akegan) [reply app/controllers/api/v1/short_form_controller.rb:22]: @software-project you're right, if we don't have the validator (and it wasn't really doing anything before) we can just remove this piece.

File: app/models/force/object_base.rb [hunk 15]
```diff
@@ -65,24 +66,48 @@ def add_salesforce_suffix(salesforce_fields)
...
       domain_fields[field_name] = ActionController::Base.helpers.number_to_currency(@fields.domain[field_name])
     end
 
     def currency_to_float(field_name, salesforce_fields)
       return nil unless @fields.salesforce[field_name]
+
       salesforce_fields[field_name] = salesforce_fields[field_name].gsub(/[$,]/, '').to_f
     end
 
+    # TODO: Automate this conversion for date objects.
```
Comment (akegan) [inline app/models/force/object_base.rb:79]: We can skip the automated date conversion in this PR and leave this TODO for a future PR

File: app/javascript/components/mappers/soqlToDomain/application_member.js [hunk 16]
```diff
@@ -11,6 +11,7 @@ export const mapApplicationMember = (value) => {
     second_phone_type: value.Second_Phone_Type,
     second_phone: value.Second_Phone,
     email: value.Email,
+    // TODO is this needed?
     primary_language: value.Primary_Language,
```
Comment (software-project) [inline app/javascript/components/mappers/soqlToDomain/application_member.js:15]: This is the second question. This param is not permitted on the supplemental application form. @akegan

File: app/javascript/components/mappers/soqlToDomain/application_member.js [hunk 17]
```diff
@@ -11,6 +11,7 @@ export const mapApplicationMember = (value) => {
     second_phone_type: value.Second_Phone_Type,
     second_phone: value.Second_Phone,
     email: value.Email,
+    // TODO is this needed?
```
Comment (akegan) [inline app/javascript/components/mappers/soqlToDomain/application_member.js:14]: I think it's probably not needed on the app member

Comment (akegan) [reply app/javascript/components/mappers/soqlToDomain/application_member.js:14]: I'll look into it more though.

File: app/controllers/api/v1/short_form_controller.rb [hunk 18]
```diff
@@ -6,159 +6,31 @@ class Api::V1::ShortFormController < ApiController
...
-        logger.debug "updated application: #{application}"
-      end
-
-      render json: { application: application }
-    else
-      render status: 422, json: { errors: short_form_validator.errors.full_messages || 'Unknown Error' }
+    # short_form_validator = ShortFormValidator.new(application_api_params)
+    # if short_form_validator.valid?
+    # FIXME: validatate / permit on the params
+    # Unpermitted parameter: :primary_language
```
Comment (akegan) [inline app/controllers/api/v1/short_form_controller.rb]: We can remove this too, and probably remove the whole short_form_validator file. under apps/services/short_form_validator.rb
```suggestion
    # TODO: Consider adding a validator to check that all required params are present
```

**Comment by akegan** [comment]
@software-project yarn unit tests are failing. 

Edit: I'll see if I can fix them

**Comment by akegan** [comment]
Fixed the unit tests - needed to update how listing info was being passed up to listings applications table. They actually caught a real bug 🎉 

We'll see if the e2e tests work now .

**Comment by akegan** [comment]
Looks like the remaining issue is with the alt contact not getting deleted properly, I'll take a look into it now

**Comment by akegan** [comment]
Tests should be fixed now. I think I'm OK with this PR if you are @software-project. let me know and we can merge it

**Comment by software-project** [comment]
@akegan I completely missed js test. Thank you for the fixes. Looks good to me.

**Review by software-project — APPROVED** [review]
👍
